tiatoolbox/models/architecture/mapde.py

```diff
@@ -0,0 +1,294 @@
+        dist_filter = np.repeat(dist_filter, repeats=num_class * 2, axis=1)
+
+        self.min_distance = min_distance
+        self.threshold_abs = threshold_abs
+        self.register_buffer(
+            "dist_filter",
+            torch.from_numpy(dist_filter.astype(np.float32)),
+        )
+        self.dist_filter.requires_grad = False
+
+    def forward(self, input_tensor: torch.Tensor) -> torch.Tensor:
+        """Logic for using layers defined in init.
+
+        This method defines how layers are used in forward operation.
+
+        Args:
+            input_tensor (torch.Tensor):
+                Input images, the tensor is in the shape of NCHW.
+
+        Returns:
+            torch.Tensor:
+                Output map for cell detection. Peak detection should be applied
+                to this output for cell detection.
+
+        """
+        logits, _, _, _ = super().forward(input_tensor)
+        out = F.conv2d(logits, self.dist_filter, padding="same")
+        return F.relu(out)
+
+    #  skipcq: PYL-W0221  # noqa: E800
+    def postproc(self, prediction_map: np.ndarray) -> np.ndarray:
+        """Post-processing script for MicroNet.
+
+        Performs peak detection and extracts coordinates in x, y format.
+
+        Args:
+            prediction_map (ndarray):
+                Input image of type numpy array.
+
+        Returns:
+            :class:`numpy.ndarray`:
+                Pixel-wise nuclear instance segmentation
+                prediction.
+
+        """
+        coordinates = peak_local_max(
+            np.squeeze(prediction_map[0], axis=2),
+            min_distance=self.min_distance,
+            threshold_abs=self.threshold_abs,
+            exclude_border=False,
+        )
+        return np.fliplr(coordinates)
+
+    @staticmethod
+    def infer_batch(model, batch_data, on_gpu):
+        """Run inference on an input batch.
+
+        This contains logic for forward operation as well as batch I/O
+        aggregation.
+
+        Args:
+            model (nn.Module):
+                PyTorch defined model.
+            batch_data (:class:`numpy.ndarray`):
+                A batch of data generated by
+                `torch.utils.data.DataLoader`.
+            on_gpu (bool):
+                Whether to run inference on a GPU.
+
+        Returns:
+            List of output from each head, each head is expected to
+            contain N predictions for N input patches. There are two
+            cases, one with 2 heads (Nuclei Pixels `np` and Hover `hv`)
+            or with 2 heads (`np`, `hv`, and Nuclei Types `tp`).
+
+        """
+        patch_imgs = batch_data
+
+        device = select_device(on_gpu)
+        patch_imgs_gpu = patch_imgs.to(device).type(torch.float32)  # to NCHW
+        patch_imgs_gpu = patch_imgs_gpu.permute(0, 3, 1, 2).contiguous()
+
+        model.eval()  # infer mode
+
+        # --------------------------------------------------------------
```

Comment:
What does this line mean? Is it to mark the start of the inference?

Reply:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/443/commits/b25b74c348f761586e459e00f51a2169b66edb8f